art/attacks/svm_attack.py

```diff
@@ -0,0 +1,264 @@
+
+
+class SVMAttack(Attack):
+    """
+    Close implementation of Biggio's attack on SVMs
+    Paper link: https://arxiv.org/pdf/1206.6389.pdf
+    """
+    attack_params = ['classifier', 'step', 'eps', 'x_train', 'y_train', 'x_val', 'y_val']
+
+    def __init__(self, classifier, step, eps, x_train, y_train, x_val=None, y_val=None, max_iters=100, **kwargs):
+        """
+        Initialize an SVM attack
+
+        :param classifier: A trained ScikitlearnSVC classifier
+        :type classifier: `art.classifiers.scikitlearn.ScikitlearnSVC`
+        :param step: The step size of the classifier
+        :type step: float
+        :param eps: The minimum difference in loss before convergence of the classifier
+        :type eps: float
+        :param x_train: The training data used for classification
+        :type x_train: `np.ndarray`
+        :param y_train: The training labels used for classification
+        :type y_train: `np.ndarray`
+        :param x_val: The validation data used to test the attack
+        :type x_val: `np.ndarray`
+        :param y_val: The validation labels used to test the attack
+        :type y_val: `np.ndarray`
+        :param max_iters: The maximum number of iterations for the attack
+        :type max_iters: int
+        :param kwargs: Extra optional keyword arguments
+        """
+        super(SVMAttack, self).__init__(classifier)
+
+        if not isinstance(classifier, ScikitlearnSVC):
+            raise TypeError('Classifier must be a SVC')
+        if isinstance(self.classifier._model, LinearSVC):
+            self.classifier = ScikitlearnSVC(model=SVC(C=self.classifier._model.C, kernel='linear'),
+                                             clip_values=self.classifier.clip_values)
+            self.classifier.fit(x_train, y_train)
+        elif not isinstance(self.classifier._model, SVC):
+            raise NotImplementedError("Model type '{}' not yet supported".format(type(self.classifier._model)))
+
+        self.step = step
+        self.eps = eps
+        self.x_train = x_train
+        self.y_train = y_train
+        self.x_val = x_val
+        self.y_val = y_val
+        self.max_iters = max_iters
+        self.set_params(**kwargs)
+
+    def generate(self, x, **kwargs):
+        """
+        Iteratively finds optimal attack points starting at values at x
+
+        :param x: An array with the points that initialize attack points.
+        :type x: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+
+        if 'y' in kwargs:
+            y_attack = kwargs['y']
+        else:
+            y_attack = self.classifier.predict(x)
+
+        num_poison = len(x)
+        if num_poison == 0:
+            raise ValueError("Must input at least one poison point")
+        num_features = len(x[0])
+        train_data = np.copy(self.x_train)
+        train_labels = np.copy(self.y_train)
+        classifier = self.classifier
```

Comment:
Is this step necessary? I think it would refer to the same instance of classifier.

Reply:
It wouldn't. New instances of the classifier are being created by `generate_attack_point`